cas/cas.c

```diff
@@ -285,6 +286,63 @@ M0_INTERNAL bool m0_crv_is_none(const struct m0_crv *crv)
 	return memcmp(crv, &M0_CRV_INIT_NONE, sizeof(*crv)) == 0;
 }
 
+M0_INTERNAL bool m0_cas_fop_is_redoable(struct m0_fop *fop)
+{
+	struct m0_fop_type *cas_fopt = &cas_put_fopt;
+	return fop->f_type == cas_fopt;
+}
+
+M0_INTERNAL int m0_cas_fop2redo(struct m0_fop *fop,
+				struct m0_dtm0_redo *redo)
+{
+	struct m0_cas_op *op = m0_fop_data(fop);
+	struct m0_buf     payload = {};
+	int               rc;
+
+	/* TODO: encode fop opcode and fop reply as well. */
+
+	rc = m0_xcode_obj_enc_to_buf(&M0_XCODE_OBJ(m0_cas_op_xc, op),
+				     &payload.b_addr, &payload.b_nob);
+	if (rc != 0)
+		return M0_ERR(rc);
+
+	rc = m0_dtm0_redo_init(redo, &op->cg_descriptor,
+			       &payload, M0_DTX0_PAYLOAD_CAS);
+	if (rc != 0)
+		m0_buf_free(&payload);
```

Comment:
payload should be free-ed regardless the `rc` condition here.
A new buf and its memory is allocated in m0_dtm0_redo_init() -> buf2bufs_copy().
The content in the `payload` has been copied to the `redo->dtr_payload.dtp_data`.

Otherwise, mem leak will happen here.

Reply:
Yes, will fix it.

Reply:
Done.